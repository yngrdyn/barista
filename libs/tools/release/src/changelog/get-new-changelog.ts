/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as conventionalChangelog from 'conventional-changelog';
import { Readable } from 'stream';

/** Generates the new changelog based on the git history */
export function getNewChangelog(
  headerPartial: string,
  releaseName: string,
): Readable {
  return conventionalChangelog(
    { preset: 'angular' },
    { title: releaseName },
    null, // raw-commits options
    {}, // commit parser options
    {
      // writer options
      headerPartial,
    },
  );
}
