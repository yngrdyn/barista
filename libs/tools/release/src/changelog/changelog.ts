/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createReadStream, createWriteStream, promises as fs } from 'fs';
import { mergeNodeStreams } from '../utils/merge-node-streams';
import { getNewChangelog } from './get-new-changelog';

/** Default filename for the changelog. */
export const CHANGELOG_FILE_NAME = 'CHANGELOG.md';

/** Writes the changelog from the latest Semver tag to the current HEAD. */
export async function prependChangelogFromLatestTag(
  changelogPath: string,
  headerPartialPath: string,
  releaseName: string = '',
): Promise<void | Error> {
  const headerPartial = await fs.readFile(headerPartialPath, {
    encoding: 'utf-8',
  });

  const newChangelogStream: any = getNewChangelog(headerPartial, releaseName);

  // Stream for reading the existing changelog. This is necessary because we want to
  // actually prepend the new changelog to the existing one.
  const previousChangelogStream = createReadStream(changelogPath);

  // Sequentially merge the changelog output and the previous changelog stream, so that
  // the new changelog section comes before the existing versions. Afterwards, pipe into the
  // changelog file, so that the changes are reflected on file system.
  const mergedCompleteChangelog = mergeNodeStreams(
    newChangelogStream,
    previousChangelogStream,
  );

  return new Promise<void | Error>((resolve, reject) => {
    // Wait for the previous changelog to be completely read because otherwise we would
    // read and write from the same source which causes the content to be thrown off.
    previousChangelogStream.on('end', () => {
      mergedCompleteChangelog
        .pipe(createWriteStream(changelogPath))
        .once('error', (error: Error) => {
          reject(error);
        })
        .once('finish', () => {
          resolve();
        });
    });
  });
}
