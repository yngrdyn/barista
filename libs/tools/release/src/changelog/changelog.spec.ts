/**
 * @license
 * Copyright 2020 Dynatrace LLC
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
jest.mock('conventional-changelog');

import { createReadStream, promises as fs } from 'fs';
import { vol } from 'memfs';
import { getFixture } from '../testing/get-fixture';
import { prependChangelogFromLatestTag } from './changelog';
import * as changelog from './get-new-changelog';

const newChangelogContent = `
## features

new content
`;

beforeEach(() => {
  vol.reset();
  vol.fromJSON({
    '/CHANGELOG.md': getFixture('CHANGELOG-minimal.md'),
    '/header.hbs': '',
    newChangelog: newChangelogContent,
  });
});

test('should prepend the changelog if there is an existing one', async () => {
  jest
    .spyOn(changelog, 'getNewChangelog')
    .mockImplementation(() => createReadStream('newChangelog'));

  await prependChangelogFromLatestTag('/CHANGELOG.md', '/header.hbs');

  const updatedChangelog = await fs.readFile('/CHANGELOG.md', {
    encoding: 'utf-8',
  });
  expect(updatedChangelog).toMatchSnapshot();
});
