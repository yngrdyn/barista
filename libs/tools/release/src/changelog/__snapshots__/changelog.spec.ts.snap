// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should prepend the changelog if there is an existing one 1`] = `
"
## features

new content
## 4.15.1 (2019-12-02)

### Bug Fixes

- **context-dialog:** Fixes positioning issues by removing the closing button
  that tightly coupled the position to the trigger
"
`;
